### DISCUSSION TITLE
expecting `throw_error_status(422, "name", "must be present")` to display some sort of message at `/create-account`.. but not seeing it

### DISCUSSION THREAD

**Discussion by jf** [jeremyevans/rodauth#410] [General]
as per subject. I am trying to add new fields to the registration form (http://rodauth.jeremyevans.net/rdoc/files/doc/guides/registration_field_rdoc.html); option 1 if it matters (add new fields to the accounts table).

The sample code has this:
```
    unless name = param_or_nil("name")
      throw_error_status(422, "name", "must be present")
    end
```

1. I have yet to customize my create-account view yet (maybe this is the issue? but if so it's not clear), and just wanted to test the `before_create_account` hook and the throw. If the standard built-in create-account view does not show any error messages, what must I do to show the error messages? Looking at Rodauth's code, I'm guessing that this has something to do with `@field_errors`... but trying to access it from my customer create-account view gives `nil`.

2. is there a method to basically only return true if param is not nil and not empty? if there's nothing built-in that's fine, but I'm just wondering. I'm hoping to reject `""` empty values as well. Yes I know I could set `required='required'` on the input field... but that's a separate discussion.

**Comment by jeremyevans** [comment]
If there are errors creating the account, the default create-account view shows them.  You can use `rodauth.field_error(rodauth.login_param)` to see if there is a field error for the login param.  What you may be missing is that `@field_errors` is a Rodauth scope instance variable, but your view is executed in Roda scope, not Rodauth scope.

There isn't a method that returns whether a submitted parameter is not nil or empty.  `param(param)` will always return a string, and `param_or_nil(param)` will return a string or nil.  You can use `param(key) != ''` to explicitly check for a non-empty param.

**Reply by jf** [reply]
thank you.

> You can use rodauth.field_error(rodauth.login_param) to see if there is a field error for the login param.

ok. How do I check for the field error for my custom field called `name` in the database? I tried `rodauth.field_error(rodauth.name_param)`... but loading the create-account form borks with
```
undefined method `name_param' for an instance of #<Class:0x0000000122bb1228> (NoMethodError)
```

**Reply by jf** [reply]
nvm, I got it. On a hunch tried `rodauth.field_error('name')`, and that worked. I assume that is the way to do it? I certainly get back the `error` I set in `set_field_error`

**Reply by jf** [reply]
> If there are errors creating the account, the default create-account view shows them.

I'm not sure... What I get is the same create-form, with no "name must be present" anywhere in sight. Not in the HTML, nor in the HTTP headers. I do get the 422... but no sight of the rest of the parameters for `throw_error_status(422, "name", "must be present")`

**Reply by jeremyevans** [reply]
If you go to http://rodauth-demo.jeremyevans.net/create-account, use the inspector to remove the required flag, and click the Create Account button, the field error is shown on the resulting page "invalid login, minimum 3 characters".

So if you aren't showing an error on your custom page, it's a problem with the custom page, not a general issue with Rodauth. You mentioned that `rodauth.field_error('name')` worked, so are you using the result on the custom page you created?

**Reply by jf** [reply]
sorry if I wasnt clear. I started out *not* creating a custom page.. but simply adding the extra column to the table, + the `before_create_account` hook with the code that throws the 422. What I'm saying is that the standard built-in page did not show any error for the missing new attribute.

I do see the error message at http://rodauth-demo.jeremyevans.net/create-account. I'm trying to do a quick glance through the code now; is that from an internal `throw_error_status` call?

EDIT: what's more important to me is actually figuring out how to detect, and then show errors. With `rodauth.field_error('name')` working, I think I have my answer.

**Reply by jf** [reply]
Just to update: I see now that the standard built-in page does show errors, but it is for the `login` field, as part of `rodauth.render('login-field')`. I was hoping to be able to have the page show errors for the *new* field ("name"); but I see now that that's something else altogether that I must build for myself.

**Comment by jf** [comment]
**UPDATE:**

I now have a custom `create-form` view, that has a `name` field (rendered using HTML), and a login field (rendered using `rodauth.render('login-field')`)

I've also got debug statements in a `before_create_account` block (as recommended by http://rodauth.jeremyevans.net/rdoc/files/doc/guides/registration_field_rdoc.html) to help validate the `name` field.

I am expecting for `before_create_account` to ALWAYS be called each time the form is submitted; but it isn't the case for me:

ONE: if I submit a string that parses as a valid email address, `before_create_account` gets called

TWO: if I submit an *invalid* email address (I use `a@a`; apparently Firefox allows me to do this), I find that the `before_create_account` never runs (no debug statements are printed). There is still some validation going on apparently (just that `before_create_account` is skipped?); because the page comes back with both `flash['error']` ("There was an error creating your account"), and `rodauth.field_error('login')` ("invalid login, not a valid email address") displayed on the page.

Perhaps I could say that I should be happy that there IS still some validaton happening... but it would appear that my own validation / `before_create_account` hook does not run UNTIL the `login` field validates. Is this by design? Even if so, I was expecting my "before" hook to run first.

**Reply by jeremyevans** [reply]
It's by design that `before_create_account` is only called if Rodauth expects it will create the account, and is not called if earlier checks fail.

If you want to run code before the other checks, use `before_create_account_route` (note that this is called for both GET and POST requests).

**Reply by jf** [reply]
ok I see. Thank you for the explanation and help!

If I may ask, how do I get access to the request object to check if it's a GET or a POST in `before_create_account_route`?

**Reply by jeremyevans** [reply]
Call the `request` method.
